### PR TITLE
add translation for lec14

### PR DESCRIPTION
**Pull request by DawnMagnet**

### CODE REVIEW

File: course14/lec14_script_en.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+Hello everyone, welcome to the modern programming ideology course brought to you by the Basic Software Center of the IDEA Research Institute.
```
Comment (peter-jerry-ye) [inline course14/lec14_script_en.md]:
```suggestion
# Case Study: Stack-based Virtual Machine
```

**Comment by DawnMagnet** [comment]
@peter-jerry-ye

File: course14/lec14_script_en.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+Now that we have mentioned WebAssembly, let's give it a brief introduction. WebAssembly, as its name suggests, Web+Assembly, is a virtual instruction set. It was initially used on the web and can be used in browsers, but since it is an instruction set, it can be executed on other platforms as long as a virtual machine is implemented, so there are also runtimes like Wasmtime, WAMR, WasmEdge, etc. It is also the first compilation backend of MoonBit. One of its major features is that its instruction set also has a type system, so there is a great guarantee of security. Today we will take a subset of the WebAssembly instruction set as an example.
```
Comment (peter-jerry-ye) [inline course14/lec14_script_en.md]: Add URL for Wasmtime/WAMR/WasmEdge?

File: course14/lec14_script_en.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+The data we consider today is only 32-bit signed integers, that is, Int in Moonbit. To meet the needs of control flow, we use non-zero integers to represent true, and zero to represent false. Our instruction set includes: data operations, data storage, and control flow. On the data side, const can define an integer constant, as well as addition, subtraction, equality judgment, and modulo. Data storage includes fetching and storing operations from local parameters. Control flow includes if-else and function calls.
+
+The definition in Moonbit is basically a one-to-one copy: data is defined using an enumeration type, although we only have 32-bit signed integers; instructions, as just introduced, include the definition of constants, addition, modulo, equality, and function calls, local fetching and setting values, all identified by strings. Finally, there is conditional judgment, which includes an integer and two instruction lists. The integer represents how many calculation results will be left after this expression block ends, and the two instruction lists correspond to the situations when the condition is true and false, respectively.
```
Comment (peter-jerry-ye) [inline course14/lec14_script_en.md]: And where's the definition?

File: course14/lec14_script_en.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+The function type definition includes the function name, input parameters, output parameters, local variables, and the function body. Here, since we only have one data type, we only record their respective names and quantities, without considering the specific corresponding types. A complete program includes multiple functions and an optional function as the program entry point.
+
+Let's take a look at the specific operation process of the stack-based virtual machine. First is numerical calculation. Taking 1+2 as an example, we have a stack, which is initially empty. The first thing we need to do is to add data to the stack. Since it is a constant, we use the constant instruction to add 1 and 2. Then, we use the addition instruction. The addition instruction will take the top two values of the stack, add them together, and store the result back on the top of the stack. So, after the operation is completed, the top element of the stack is 3.
```
Comment (peter-jerry-ye) [inline course14/lec14_script_en.md:36]: Don't forget to put all the illustrations

File: course14/lec14_en.md [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+
+  ```moonbit
+  enum Expression {
+  Number(Int)
+  Plus(Expression, Expression)
+  Minus(Expression, Expression)
+  Multiply(Expression, Expression)
+  Divide(Expression, Expression)
+  }
+  ```
```
Comment (peter-jerry-ye) [inline course14/lec14_en.md]:
```suggestion
  ```moonbit
  enum Expression {
    Number(Int)
    Plus(Expression, Expression)
    Minus(Expression, Expression)
    Multiply(Expression, Expression)
    Divide(Expression, Expression)
  }
  ```
```

File: course14/lec14_en.md [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+- Utilize the built-in Buffer data structure for more efficient operation than string concatenation.
+
+  ```moonbit
+  fn Instruction::to_wasm(self : Instruction, buffer : Buffer) -> Unit {
+  match self {
+      Add => buffer.write_string("i32.add ")
+      Local_Get(val) => buffer.write_string("local.get $" + val + " ")
+      _ => buffer.write_string("...")
+  }
+  }
```
Comment (peter-jerry-ye) [inline course14/lec14_en.md]:
```suggestion
  fn Instruction::to_wasm(self : Instruction, buffer : Buffer) -> Unit {
    match self {
      Add => buffer.write_string("i32.add ")
      Local_Get(val) => buffer.write_string("local.get $" + val + " ")
      _ => buffer.write_string("...")
    }
  }
```

File: course14/lec14_en.md [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+
+  ```moonbit no-check
+  fn compile_expression(expression : Expression) -> List[Instruction] {
+  match expression {
+      Number(i) => List::[Const(I32(i))]
+      Plus(a, b) => compile_expression(a) + compile_expression(b) + List::[Add]
+      Minus(a, b) => compile_expression(a) + compile_expression(b) + List::[Sub]
+      _ => List::[]
+  }
+  }
```
Comment (peter-jerry-ye) [inline course14/lec14_en.md]:
```suggestion
  fn compile_expression(expression : Expression) -> List[Instruction] {
    match expression {
      Number(i) => List::[Const(I32(i))]
      Plus(a, b) => compile_expression(a) + compile_expression(b) + List::[Add]
      Minus(a, b) => compile_expression(a) + compile_expression(b) + List::[Sub]
      _ => List::[]
    }
  }
```

**Review by peter-jerry-ye — CHANGES_REQUESTED** [review]

**Pull request merged**